Added hierarchical/nested tags

vsimkus opened the pull request:
Potentially resolves 
* https://github.com/laurent22/joplin/issues/375
* https://discourse.joplinapp.org/t/nested-tags-nested-notebook-structure-from-evernote/1562.

## Details

The implementation uses `/` symbol as a nesting separator. I.e. `tag/subtag` is a nested tag, where `tag` is the parent tag and `subtag` is its child. Creating a tag named `tag/subtag/subsubtag` creates three tags, one for each level. The tags are associated using `parent_id` field.

In the app, viewing notes with a tag will also show all notes that are associated with any of the tag's descendant tags (same for the note count). Deleting a tag will also delete all its descendant tags.

In the desktop app the tags are shown nested just like the notebooks:
![Screenshot from 2020-02-24 16-23-08](https://user-images.githubusercontent.com/5730052/75165085-1f120280-5722-11ea-92d4-01b97c9f3fff.png)
And in the editor they are shown using the `/` separator:
![Screenshot from 2020-02-24 16-25-22](https://user-images.githubusercontent.com/5730052/75165216-55e81880-5722-11ea-8c2b-a54b91d96857.png)

The tag hierarchy can be changed by drag-and-drop. I.e. you can drag a tag and drop it into the default scope or place it under another tag.

## Examples
1. Creating a new tag `a/b` creates a tag `a/b` and `a` (if it does not already exist). It then attaches `a/b` to the current note. 
In the tag explorer, the note will show up whenever `a` or `a/b` is selected. 
Also, the note will add to the count of notes for `a` and `a/b`. 
2. Dragging a tag `a/b` onto the `Tags` section will rename the tag to `b`, essentially moving it from hierarchy under `a` to the default scope.
3. Dragging a tag `a/b/c` onto `a` will create `a/c`. The note count for `a/b` is automatically updated.
4. If `b` already exists, and you try to drag `a/b` onto the default scope, then the error popup will be shown `The tag "%s" already exists. Please choose a different name.`.
5. The same dragging behaviour exists when renaming tags. Additionally, if you rename `a/b` to `c/b` and `c` does not already exist, then the parent tag `c` will be created.
6. Deleting a tag `a`, when it has descendants `a/b` and `a/c` will delete all tags `a`, `a/b`, and `a/c` - the deletion confirmation popup reminds that with `Remove tag "%s" and its descendant tags from all notes?'`.

laurent22 commented:
Thanks for the pull request @vsimkus. It's quite big so I will need some time to review it.

vsimkus commented:
@laurent22 No worries, take your time. Let me know of any suggestions to improve it.
I think one potential improvement is to create a view table that maps parent tag ids to their descendants. Such a mapping is currently used in `tags_with_note_count` view and `noteIds()` method. But I'd like your input before creating additional tables/views.

Also, I've noticed another bug related related to renaming of tags and reported it https://github.com/laurent22/joplin/issues/2618. This bug is not fixed in this PR.

tessus commented:
I have a question about the separator:

What happens, if there's a tag with a `/` in it? I'd say `:` would be a better separator, but then we still had the same problem. How are things like this usually solved?

Maybe a unicode symbol, e.g.:

- middle dot `·` (U+00B7) 
- bullet `•`  (U+2022)
- vertical ellipsis `⋮` (U+22EE)

vsimkus commented:
@tessus The implementation in this PR won't allow `/` in the tag titles. Creating a tag with a `/` in it's name splits it into parent/descendant tags, i.e. `tag/subtag/subsubtag` actually creates three tags `tag`, `tag/subtag`, and `tag/subtag/subsubtag`, where the display title (the one that is shown on the sidebar) of a tag is the part after the last `/`.  Typing any of the atypical symbols wouldn't be convenient to the user.

This has a few advantages:
1. It makes the creation of nested tags easy - you just create a tag with name delimited by `/` e.g. `tag/something`, and it handles the creation of the parent tags.
2. You can change the hierarchy of the tag by simply renaming it. E.g. if you wanted to move a tag `something` from `tag/something` hierarchy under `tag2/subtag` you could just rename it from `tag/something` to `tag2/subtag/something`.

The alternative to this, is to make nested tag creation similar to nested notebooks, where you initially create the tag with the default empty parent. Then move it under another tag, i.e. change the parent of the tag. This way you wouldn't need to reserve any character for splitting tags. IMO, it is a bit less convenient since it requires more work when creating new tags.

As to your other question
> How are things like this usually solved?

Here's a discussion in Evernote forum: https://discussion.evernote.com/topic/118455-true-hierarchical-tags/. In Notable and Bearnote the nested tags are created similarly by using a `/` delimiter.
As suggested in the link above, we could let the users customize the delimiter used by adding a setting, but I doubt it would be used much.

tessus commented:
Thanks for the explanation. I certainly do understand the advantages, but was worried that people would complain, if they couldn't use  a slash in the tag name.

But apparently they are already used to that from Evernote, so all is good. 👍

coderrsid commented:
Hey @laurent22  @tessus  I wanted to ask that if #375 is resolved by this PR but it's still in GSOC ideas list. So can i still draft a proposal for that?

laurent22 commented:
@coderrsid, there's already a PR for it: https://github.com/laurent22/joplin/pull/2572

laurent22 commented on ReactNativeClient/lib/joplin-database.js:
```diff
@@ -686,6 +686,24 @@ class JoplinDatabase extends Database {
 				queries.push('ALTER TABLE version ADD COLUMN table_fields_version INT NOT NULL DEFAULT 0');
 			}
 
+			if (targetVersion == 30) {
+				queries.push('ALTER TABLE tags ADD COLUMN parent_id TEXT NOT NULL DEFAULT ""');
+				// Modify the tags_with_note_count view table for hierarchical tags
+				queries.push('DROP VIEW tags_with_note_count');
+				queries.push(`CREATE VIEW tags_with_note_count AS
+									WITH RECURSIVE
+									parent_of(id, parent_id, title, created_time, updated_time, child_id) AS 
+									(SELECT id, parent_id, title, created_time, updated_time, id FROM tags
+										UNION ALL
+										SELECT parent_of.id, parent_of.parent_id, parent_of.title, parent_of.created_time, parent_of.updated_time, tags2.id FROM parent_of JOIN tags AS tags2 ON parent_of.child_id=tags2.parent_id)
+									SELECT parent_of.id, parent_of.parent_id, parent_of.title, parent_of.created_time, parent_of.updated_time, COUNT(DISTINCT notes.id) as note_count FROM
+											parent_of
+											LEFT JOIN note_tags nt on nt.tag_id = parent_of.child_id 
+											LEFT JOIN notes on notes.id = nt.note_id 
+											WHERE notes.id IS NOT NULL
+										GROUP BY parent_of.id`);
+			}
+
```
I'm not really keen on this big SQL query for the following reasons:
- It's huge and we don't want to encode too much business logic in the database
- Every time a field is added to the tag table, it should be duplicated here
- It adds a non-existing "note_count" field to the tag. It means if somewhere in the code something tries to save this tag it will fail. In general I've been quite strict about making sure that there's a one-to-one mapping between table fields and model fields so that you don't need to add extra code and checks before saving a model, and I'd prefer to keep it that way.

For folders we have a `Folder.addNoteCounts()` method, which I don't like much either, but at least it makes it clear from the calling code that we are somehow adding something to the list of folders.

But ideally there should be a separate structure, an associate array that maps a tag ID to a note count. And that structure would be updated whenever relevant. That way it's easy to access the note count, and the actual tag models are not modified.

I'm not sure how easy it is to implement though. Perhaps the note count could be removed from this pull request to keep things simple, and a second pull request could add that back.

vsimkus replied:
The reason I've kept the SQL query, was due to the way that note counting was done till now, which used a SQL query defined at `targetVersion == 25`. Obviously with recursion it became rather bulky, so I agree that it would be best to factor it out into its own model. I'll have a look into the effort required, or else to prevent the complication of this PR I will remove the note counting from this PR.

vsimkus replied:
Ok so for the moment I've removed the `tags_with_note_count` view table and implemented an equivalent function `Tag.addNoteCounts()` that counts the number of linked notes on the fly and adds the attribute to the object. Adding the function is a bit less overarching than (temporarily) removing the tag note count entirely. 

As to creating a separate structure for tag note count - it is also a bit more complex. The reason is that there is (at the moment) nothing stopping the user from tagging the same note with two tags from the same hierarchy e.g. `a/b` and `a/b/c`, in which case when incrementing the note count for a specific tag you would need to know if the note has already been associated with another tag from the same hierarchy. You could potentially solve this by maintaining a separate associative structure `tag_id->note_id`, which would link a tag with any of its descendant tags' notes.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -14,7 +14,13 @@ class Tag extends BaseItem {
 	static async noteIds(tagId) {
-		const rows = await this.db().selectAll('SELECT note_id FROM note_tags WHERE tag_id = ?', [tagId]);
+		// Get NoteIds of that are tagged with current tag or its descendants
+		const rows = await this.db().selectAll(`WITH RECURSIVE
+												parent_of(id, child_id) AS 
+												(SELECT id, id FROM tags where id=?
+												UNION ALL
+												SELECT parent_of.id, tags2.id FROM parent_of JOIN tags AS tags2 ON parent_of.child_id=tags2.parent_id)
+												SELECT note_id FROM note_tags WHERE tag_id IN (SELECT child_id from parent_of)`, [tagId]);
```
Again I'd prefer if this is done in JavaScript rather than in SQL.

vsimkus replied:
Moved this to JS.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -36,18 +42,48 @@ class Tag extends BaseItem {
 
-		await Tag.delete(tagId);
+		// We're recursing over all children already, so this call doesn't have to be recursive.
+		await Tag.delete(tagId, { deleteChildren: false });
```
The purpose of this `deleteChildren` option is just to save a call to `Tag.childrenTagIds`? In that case, I think it's ok to remove it, as we probably won't get much better performance from it.

vsimkus replied:
Yes, that's correct.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -36,18 +42,48 @@ class Tag extends BaseItem {
 
 	static async delete(id, options = null) {
 		if (!options) options = {};
+		if (!('deleteChildren' in options)) options.deleteChildren = true;
+
+		const tag = await Tag.load(id);
+		if (!tag) return; // noop
+
+		if (options.deleteChildren) {
+			const childrenTagIds = await Tag.childrenTagIds(id);
+			for (let i = 0; i < childrenTagIds.length; i++) {
+				await Tag.delete(childrenTagIds[i]);
+			}
+		}
```
Are there cases where we don't want to delete the children? If not, I think we can also remove the option from here.

laurent22 replied:
Actually the purpose of `deleteChildren` in Folder.delete is to prevent it from happening in the synchronizer. Because in general synchronizer actions should not have side-effects or it can lead to very strange issue, synchronisation loops, etc. So if you add it there, you should also use that option where relevant in synchronizer.js

vsimkus replied:
That makes sense. I've created the `deleteChildren` option to match the interface of `Folder.delete` since they are alike. I missed its use in `synchronizer.js` - thanks, I'll look into it.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -174,14 +210,121 @@ class Tag extends BaseItem {
 		}
 	}
 
+	static tagPath(tags, tagId) {
+		const idToTags = {};
+		for (let i = 0; i < tags.length; i++) {
+			idToTags[tags[i].id] = tags[i];
+		}
+
+		const path = [];
+		while (tagId) {
+			const tag = idToTags[tagId];
+			if (!tag) break; // Shouldn't happen
+			path.push(tag);
+			tagId = tag.parent_id;
+		}
+
+		path.reverse();
+
+		return path;
+	}
```
Please create a utility method shared by Folder.folderPath and tag.tagPath

vsimkus replied:
I've created a `nested-utils.js` file, since I wasn't sure where else that utility could go.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -174,14 +210,121 @@ class Tag extends BaseItem {
+
+		if ('title' in o) {
+			o.title = o.title.trim().toLowerCase();
+			const existingTagByTitle = await Tag.loadByTitle(o.title);
+
+			if (options && options.userSideValidation && existingTagByTitle && existingTagByTitle.id !== o.id) {
+				throw new Error(_('The tag "%s" already exists. Please choose a different name.', o.title));
+			}
+
+			let parentId = '';
+			let parentTitle = '';
+			// Check if the tag is nested using `/` as separator
+			const separator = '/';
+			const i = o.title.lastIndexOf(separator);
+			if (i !== -1) {
+				parentTitle = o.title.slice(0,i);
+
+				// Try to get the parent tag
+				const parentTag = await Tag.loadByTitle(parentTitle);
+				if (parentTag) parentId = parentTag.id;
+			}
 
-				const existingTag = await Tag.loadByTitle(o.title);
-				if (existingTag && existingTag.id !== o.id) throw new Error(_('The tag "%s" already exists. Please choose a different name.', o.title));
+			if (options && options.createParents && parentTitle && !parentId) {
+				// Create the parent tag if it doesn't exist
+				const parent = {
+					title: parentTitle,
+				};
+				// Do not select the parent note_tags
+				const parentOpts = {};
+				parentOpts.selectNoteTag = false;
+				const parentTag = await Tag.save(parent, parentOpts);
+				parentId = parentTag.id;
 			}
+			// Set parent_id
+			o.parent_id = parentId;
```
The `save()` method is going to be used during sync so the data shouldn't be changed at this point as it can create strange inconsistency. It should not load data either as that data may or may not be there during the sync process. In general, this kind of business logic should be done before, when it's needed, and not when saving.

vsimkus replied:
Thanks, I do need to look into the synchronizer, and will then get this fixed.

vsimkus replied:
Created a `Tag.saveNested()` method, that handles nested tag creation and kept the `save()` method untouched.

vsimkus commented on ElectronClient/gui/SideBar.jsx:
```diff
@@ -74,6 +88,17 @@ class SideBarComponent extends React.Component {
 				for (let i = 0; i < noteIds.length; i++) {
 					await Tag.addNote(tagId, noteIds[i]);
 				}
+			} else if (dt.types.indexOf('text/x-jop-tag-ids') >= 0) {
+				event.preventDefault();
+
+				const tagIds = JSON.parse(dt.getData('text/x-jop-tag-ids'));
+				try {
+					for (let i = 0; i < tagIds.length; i++) {
+						await Tag.moveTag(tagIds[i], tagId);
+					}
+				} catch (error) {
+					bridge().showErrorMessageBox(error.message);
+				}
```
Thanks for spotting this, it should be easy to fix.

vsimkus commented:
@laurent22 Thanks for reviewing, I will make the changes as soon as I can, although I am a little short on time at the moment so it might take a while - I will look into this or next weekend.

JeffRockatansky commented:
@vsimkus any updates on this pull?

vsimkus commented:
The PR is mostly done but a couple of things are still left to be done: 
* Add a fail-safe in case `tag.id == tag.parent_id`, although that is supposed to be prevented by the safety checks.
* Change the synchroniser such that the deletion of a tag doesn't have side effects - I haven't investigated this at all so far.
* Add a couple more tests

I'll probably do this in the coming weekend if I find some spare time.

stloewen commented:
I already use tags containing a `/` to denote a hierarchy so I'm really looking forward to this feature.
Will existing tags with a `/` be handled correctly (i.e. split into hierarchical levels) when this feature is introduced or will it lead to clashes?

laurent22 commented:
@vsimkus, thanks for looking into these changes. Additionally, I think we should indeed define what happens to already existing tags that contain "/". Either we leave them as they are, or we split them into a hierarchy, but it needs to be clear so that it's not handled in one way in one part of the code and in a different way in another part.

vsimkus commented:
@laurent22 I think it would be best to convert the existing tags with `/` to nested ones somewhere during the upgrade. At the moment, for such tags only the last segment after the final `/` would be displayed in the left-side panel.
 
Is there a common place for such upgrade scripts? Something like `joplin-database.js` but for arbitrary migration code? I could of course do the migration using a sql script, but it would be easier to reuse existing javascript code. I think it should be sufficient to "save" the existing tags again using `nestedSave` and that should create the parent tags where necessary.

laurent22 commented:
To add a JS-based migration, you can use the function `addMigration(VERSION_NUM)` in joplin-database.js. Then place the file in `ReactNativeClient/lib/migrations/VERSION_NUM.js`.

I agree with your suggestion to convert the existing tags with `/`. We also need to decide how to handle cases where it would create a tag that already exist.

For example, if a user has a tag `photos/france` and also a tag `france`. In that case, we shouldn't merge the tag, but create two separate one. I guess the app already handles tags with the same name, but maybe if they are on the same level you could rename the new one to "france-1" for example.

laurent22 commented:
Also when splitting a tag, I think the leaf one should simply be renamed (and not newly created). For example, for "photos/france", you would create a new "photos" tag and then you would rename "photos/france" to "france". That way, existing notes for that tag will stay where they are. What do you think?

vsimkus commented:
Thanks for the pointers for the migration script.

As far as I can tell, currently Joplin does not allow two tags with the same name (handled by the `userSideValidation` option in `Tag.save()`). Now with nested tags the same validation applies - a user cannot create two tags at the same level with the same name. However, creating any number of tags e.g. `photos/france`, `france`, `europe/countries/france` is allowed. 
This is because all tag titles in the database are their fully qualified titles e.g. `photos/france`. In contrast to storing only the display name (segment after the last `/`) this simplifies many things in terms of tag-search and display. I strip the segment up to and including the last `/` when displaying the tags in the left-side panel. This could also be approached by storing another column in the table `display_title`, but not sure if that would be needed since the way we display the tag varies depending on the component - in the editor and search we display them using the full names (e.g. `photos/france`) and in the left-hand panel we display only the last segment i.e. for a tag `photos/france` we display `france` under a parent tag `photos`. So its better to store the full name and display it in a component-specific way with some extra code.

Then, in terms of migration I do not need to rename a tag `photos/france` to `france`, all I need is to create the parent `photos` and set its id as `parent_id` to the `photos/france` tag. This should be handled by the logic in `saveNested`, which creates parents if they do not exist, and assign them to their id as `parent_id` to their children.

vsimkus commented:
@laurent22 I've now implemented the rest of the changes and added a couple of more tests. Feel free to review again when you've got time.

Also, added checks to disallow tag titles that start/end with a `/` or have a sequence of `/` somewhere in them as that would not show them correctly and I don't think there is a legitimate use case for such tags. Creation of a disallowed tag throws an error in a window.

Finally, added a migration script that converts tags with `/` to nested tags. It also cleans up the tags if necessary - i.e. removes any sequence of starting or ending delimiters e.g. `///tag///` -> `tag`, replaces sequences of delimiters inside tags with a single delimiter e.g. `tag//tag2` -> `tag/tag2`. And in case of title clash, which might happen due to the cleanup, it checks if a tag with such a name exists, and if so adds a counter as you suggested, e.g. if you had `tag/` and `tag`, one of them will keep the name `tag` and the other will be renamed to `tag-1` (or higher counter if there are multiple clashes).

Here's an illustration of the tags before and after migration:
![migration](https://user-images.githubusercontent.com/5730052/82736688-83ddf880-9d23-11ea-8d8f-a6e38d956af9.png)

@stloewen I think this answers your question now, thanks for bringing it up.

laurent22 commented:
Thanks for the update @vsimkus, I'll take a look. 

> This is because all tag titles in the database are their fully qualified titles e.g. photos/france.

I hadn't realised this and I'm a bit concerned about it. It works well with the local database but not so much with sync. For example in these cases:

- Rename root tag that has 1000 descendants - now 1001 tags need to be synced

- Rename tag called "root" to "root2" with sub tag "root/one", "root/two". Now sync. Then if sync is interrupted, other clients will have tags like "root2/one" and "root/two". Maybe the root tag would still be called "root" as well.

Basically with sync any redundant data (in this case parent_id and title contain redundant data) will cause weird issues and inconsistencies in the data. So it's best not to have any duplicate data and to reduce to a minimum the number of changed objects in each operation.

So I think if you remove the full title from the "title" property that would make things more efficient and safer in terms of sync.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -174,6 +256,110 @@ class Tag extends BaseItem {
+		}
+	}
+
+	static async rename(tag, newTitle) {
+		const oldParentId = tag.parent_id;
+
+		tag.title = newTitle;
+		tag = await Tag.saveNested(tag, { fields: ['title', 'parent_id'], userSideValidation: true });
+		await Tag.changeDescendantPrefix(tag, newTitle);
+
+		if (oldParentId !== tag.parent_id) {
+			// If the parent tag has changed, and the ancestor doesn't
+			// have notes attached, then remove it
+			const oldParentWithCount = await Tag.loadWithCount(oldParentId);
+			if (!oldParentWithCount) {
+				await Tag.delete(oldParentId, { deleteChildren: false, deleteNotelessParents: true });
+			}
+		}
+		return tag;
+	}
+
+	static async moveTag(tagId, parentTagId) {
+		if (tagId === parentTagId
+				|| (await Tag.descendantTagIds(tagId)).includes(parentTagId)) {
+			throw new Error(_('Cannot move tag to this location.'));
```
There shouldn't be special methods to change tag properties because they can be modified via the API too for example. It should be possible to call `Tag.save({ title: "abc" })` or `Tag.save({ parent_id: "123" })` and to be confident the app is going to do the right thing. So if there's extra logic that need to happen in this case, it should be moved to Tag.save (preferably under `underSideValidation`.

Also please append a `_` to private methods. You can perhaps use this as private helper methods, that you'd call from Tag.save.

vsimkus replied:
The purpose of this method is slightly different now after the latest change - it is used to rename the full_title of a tag, i.e. `a/b/c`->`b/c/d`, it handles the creation of new parent tags (if necessary) and deletes old parents if they're no longer linked to any notes. I've changed the name to `renameNested` to reflect this. 
You can still change the short `title` and `parent_id` using the `Tag.save` method.

vsimkus commented:
@laurent22 I would imagine cases like these where users have thousands of tags might be very rare as it makes the whole point of organisation go away. However, I do agree on the concern where a failed sync can cause horrible inconsistencies -- I initially thought that synchronisation was an atomic action.

I've made changes to store only the short title in the DB field and I add a new full_title (virtual) field when the tags are loaded. This is of course opposite to having a one-to-one mapping between the db fields and the model fields, however I think that is the correct compromise in this case. Otherwise loading the full title on-demand will be even more overarching change and also less efficient from user perspective.

laurent22 commented:
Hmm, yes as mentioned earlier, we need to preserve this one-to-one mapping so adding new virtual properties unfortunately is not an option. Would it be such a large change to switch to titles with no slashes? The internal code should only care about the parent_id anyway, and it's when displaying it in some places that we might want to add the slashes. I'd expect the code will actually be simpler since there's no longer a need to parse and concat the full tag title.

Performance is not an issue, as that can always be optimised later on. However if we introduce virtual properties, it's pretty much permanent and hard to maintain over time.

vsimkus commented:
@laurent22 Thanks for the reply.

> Would it be such a large change to switch to titles with no slashes?

As mentioned above the titles are now stored without slashes, for example, a tag `a/b/c` is stored with title `c` and parent_id as the id of `a/b`. It is only when loading the tag that the full title `a/b/c` is evaluated using parent_ids and added to the tag object as `full_title` field. I suppose you didn't mean otherwise, but I wanted to clarify just in case, that the property is not stored with slashes, the slashes are added to join the titles of a tags ancestors.

> I'd expect the code will actually be simpler since there's no longer a need to parse and concat the full tag title.

Making a change such that the full_title is computed only when it is needed wouldn't be a huge change. However, it would overarch almost all uses of tags, meaning that the change would affect almost all uses of tags everywhere in the application. Let me give examples:

1. Displaying title in most cases requires the full title, that is, at the top of the editor in selected note-tags, in all (dropdown) lists (add/remove tags desktop/mobile, webclipper, tags component in mobile (this last one may change later when collapsing of tag hierarchies is added to mobile component, but currently tags are shown by their full title.)).
2. CLI autocompletion: you need to know the full_title as a tag is not anymore fully qualified by only its title, but rather by its full title.
3. Renaming tags: you generally need to display the full title so that the user can modify it. Of course you could just let the user rename only the tag's short title, i.e. `a/b/c` -> `a/b/d`, but not `a/b/c`->`a/c` or `a/b/c`->`a/b/d/c` (I am showing the full title in these examples to show that you _wouldn't_ be able change the hierarchy of the particular tag by renaming it). In case you don't allow full-title renaming, then the CLI client needs to add new commands for moving a tag from one hierarchy to another (with the current implementation this can be done by simply renaming the tag). However, it is much more convenient if you allow the user to change the full-title `a/b/c`->`a/b/1/2/3/c` for example, thus allowing to change the hierarchy of a tag by renaming it, and for such renaming you usually want to display the full title as an "initial value" that the user then modifies.
4. Searching tags in "go to anything" - you generally will want to be able to search by full title (and also display the full title of the results to be able to distinguish them).

If you do choose to compute the full-title on demand, wherever it is needed, then you will need to add this additional code almost everywhere in the application, thus spreading the feature-specific code throughout the code. The code will definitely _not_ be any simpler, since the code for computing the full title still needs to be there (and it's actually the simple bit). However, larger chucks of feature-specific code will be spread across the app.

My advice and preference would be to try and keep most of the nesting/hierarchy specific code in the `Tag.js` instead of spreading it across the app unless really necessary. It is easier to deal with a simple virtual field, whose most of the logic remains in one file, than adding complex feature-specific logic to every other component of the app. I think it will be easier from maintenance perspective.

That being said the change between the two solutions (computing full-title only when need to display and computing it when loading a tag) is quite simple, since most of the locations in the code have been identified in this PR already. 

Apologies for the long comment and thanks for you comments, let me know what you think.

laurent22 commented:
Thanks for the explanation and it makes sense as there was a few things I indeed hadn't thought about.

Overall your implementation is very good and well supported by the test units. My concern however is that this "full_title" property is really like cached data and we're attaching that to a database row, so in theory it's not great, and in practice it will cause problems now and then.

I had a look again at your code and I think it could be tweaked so as to keep the cached data separate from the database row. You could have, like now, a function that generate the full title for all the tags, but instead of attaching that data to the row, you could have a separate structure that maps a tag ID to a full title. That structure could be part of the Tag class and then you could have functions like `Tag.fullTitle(tagId)` to retrieve this data. 

I think that won't require many changes. The main issue being to make sure the structure is kept up-to-date when tag titles are changes or tags delete. What do you think?

vsimkus commented:
Indeed, I had the same concern initially (that I forgot about later). It doesn't currently cause issues, because the middleware in `reduxSharedMiddleware.js` reloads and refreshes all tags whenever any one of them is updated. I haven't noticed any inconsistencies so far.

Keeping the full_title separate makes sense and I will look into it, I suppose it shouldn't be much work to hook onto the existing events in the middleware.

vsimkus commented:
@laurent22 I've added a `fullTitleCache` in Tag.js, and add a getter for `full_title` field, which references the cache, to the tag objects when loaded. The cache is updated whenever:

1. `Tag.all` is called, since it essentially allows us to update the cache almost for free.
2. `Tag.save` is called. In this case, the cache items that refer to the current tag, or any of its descendants is updated. That is because modifying a tag's title or parent_id also requires updating the full_title cache item for its descendants.

Using a getter method for the `full_title` gives us the same functionality as calling `Tag.getCachedFullTitle(tag)` when necessary, but it is a bit cleaner I think. For example, there is no need to modify the API service for serializing the `full_title` field.

P.S. For some reason I can't trigger the CI build although I've tried pushing several times. I suppose you could set it off manually?

laurent22 commented:
@vsimkus, when I wrote this:

> You could have, like now, a function that generate the full title for all the tags, but instead of attaching that data to the row, you could have a separate structure that maps a tag ID to a full title.

What I meant is that the "full_title" should not be part of the tag, i.e. there should not be `tag.full_title`. Instead you use a function to get the full title wherever it is needed.

In practice, it means replacing all the `tag.full_title` by something like `Tag.fullTitle(tag.id)`. I assume a global search and replace should work since the full title map is kept up-to-date by your recent changes. Does that make sense?

So to be clear, this is a valid tag:

```
{ id: "123", title: "my-tag" }
```

This is not:

```
{ id: "123", title: "my-tag", full_title: "/path/to/my-tag" }
```

because "full_title" doesn't exist in the database. If it's not in the db, it should not be in the model object either.

vsimkus commented:
@laurent22 I understood what you meant and thanks for explanation again, apologies for making you do it again. The implementation you want is easy to get to but I wanted to suggest the current implementation as an alternative and discuss it first before going the other route.

My reasoning why `full_title` could be an attribute of the tag object is that it is essentially a composite read-only property of a tag composed of its own title and its parent's `full_title`. Indeed you cannot modify it directly like the other fields in the model, but I don't see a problem with having a property that is read-only. In your suggestion `Tag.fullTitle(tag.id)` is a no-side-effect "get" call to retrieve a field that is uniquely associated with the particular tag, which I think also qualifies the property as a field in the model, even if it is read-only. It's a little odd indirection to call a static class method to retrieve a property that uniquely belongs to the object.

One reason from the code side is that treating it as a property handles object serialization without additional code. Now if you want to use `Tag.fullTitle(tag.id)`, then you will need to add specific handling in `rest/Api.js`'s `defaultAction_()` (for example) to manually serialize and send the `full_title` property whenever a tag is serialized for a `GET` request. I have currently avoided this by making `full_title` a property of the object, which is required for example in the Clipper, but I suppose it might be needed in other projects such as [joplin-web](https://github.com/foxmask/joplin-web) (I know it's not an official one, so might not be quite relevant, but the at least the clipper is relevant).

Let me know what you think.

---

Actually, having written that and looked back at the code, I realised that the `Api.js` still needs a code change to handle `PUT` and `POST` for tag requests such that nested tags can be created and modified by renaming through the api.
But either way, I don't see having `full_title` as a property of an object as an evil, so let me know what you think and I will implement it to accommodate that :)

laurent22 commented:
I see what you mean but I don't think the full_title uniquely belongs to the object, as it's in fact a compound of the object's title, its parent's title, grand-parent's title, etc. Why should something that's part of another tag be part of the current tag. In the same way we don't for example have properties like "note_folder_title" on notes because that can be computed when needed. It would also be convenient to have this property I guess but for now we don't.

The second issue is that Joplin's codebase does not generally expect virtual properties on models. There's nothing to handle this or to mark a property as virtual or to track and update them, so save and load functions, serialization, etc. all expect "clean" models that exactly match what's in the database. In fact, the full_title property was accidentally synced, all clients would fail on that unexpected property (it's designed to fail hard on unexpected data to avoid corruption).

If we do support virtual properties eventually, I feel that would be a rather major feature and probably something for another pull request.

That's why in the meantime a helper function that gives you what you need might be better, so if you could change it to only use Tag.fullTitle() that would be great.

For the clipper, I think we have two options:
- Send the regular tags, and have the clipper compute the full title
- Just send the full_title as a virtual property
In this narrow case, I think it makes sense to send the full_title property as it keeps the clipper code lighter and that way we don't duplicate business logic. So I would suggest this:

- **GET /tags**: By default, make it return the tags *without* full_title. But if the "fields" parameter is present and has the "full_title" inside, then return it. So for example `GET /tags?fields=id,full_title` would return the ID and full_title.

- **POST /notes**: When posting a note, add the `tag` property like now, but send the full_title instead of the regular title. Then you will need to process this full title in `Api::action_notes`.

I think the clipper will need to be updated the get and send this full_title so if you need some help with this let me know.

vsimkus commented:
Thanks for taking the time to explain. I still think that from the modelling side, the full_title could be a property of a tag - it is as unique to one tag as its own id (it's not the same with say note folder names as you suggested), the fact that it is a composite property does not matter, once it's composed it belongs to the one tag only. But I appreciate you wanting to keep the codebase consistent, and especially if it could break Joplin somewhere down the line (although afaik the current codebase already has some virtual properties). 

I believe there could be a place for virtual properties in Joplin, as you say, and that could help clean up the codebase a little in the future, or as an effort to increase efficiency. Either way, it is actually in a very good state anyway for a project of its size and indeed has some great design decisions!

I'll make the change sometime this weekend, hopefully, if I find the time.

vsimkus commented:
@laurent22 I've removed the full_title property and now retrieve it from the cache where necessary. Also added the API for getting the full_title with the tags as you've suggested. Tested Desktop app, Mobile app and clipper. Should be good to go now. Let me know what you think.

laurent22 commented on ElectronClient/plugins/GotoAnything.jsx:
```diff
@@ -191,8 +191,11 @@ class Dialog extends React.PureComponent {
+				searchQuery = this.state.query.split(' ')[0].substr(1).trim();
+				results = await Tag.search({ fullTitleRegex: `.*${searchQuery}.*` });
+				for (let i = 0; i < results.length; i++) {
+					results[i].title = Tag.getCachedFullTitle(results[i].id);
```
Please don't mutate data, as it could cause hard to debug issues. I've noticed a few instances of this, so I'll mark them below. In this case, using `map()` to modify the array, and `Object.assign` to modify the objects might be best:
`results = results.map(tag => Object.assign({}, tag, { title: Tag.getCachedFullTitle(tag.id) })`

laurent22 commented on ReactNativeClient/lib/components/screens/notes.js:
```diff
@@ -184,6 +184,7 @@ class NotesScreenComponent extends BaseScreenComponent {
 			output = Folder.byId(props.folders, props.selectedFolderId);
 		} else if (props.notesParentType == 'Tag') {
 			output = Tag.byId(props.tags, props.selectedTagId);
+			output.title = Tag.getCachedFullTitle(output.id);
```
Please don't mutate data.

laurent22 commented on ReactNativeClient/lib/migrations/31.js:
```diff
@@ -0,0 +1,33 @@
+
+	// In case tags with `/` exist, we want to transform them into nested tags
+	for (let i = 0; i < tags.length; i++) {
+		const tag = tags[i];
```
Please don't mutate data - it could be `const tag = Object.assign({}, tags[i])`

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -86,9 +175,57 @@ class Tag extends BaseItem {
+			const noteIds = await Tag.noteIds(tag.id);
+			// Make sure the notes exist
+			const notes = await Note.byIds(noteIds);
+			tag.note_count = notes.length;
```
I think this is the same issue as the "full_title" property, and there should be a separate function to get the note count so as not to modify database rows.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -86,9 +175,57 @@ class Tag extends BaseItem {
+
+		const noteIds = await Tag.noteIds(tagId);
+		// Make sure the notes exist
+		tag.note_count = (await Note.byIds(noteIds)).length;
+		if (tag.note_count === 0) return;
```
Same here, database rows shouldn't be modified.

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -174,23 +342,131 @@ class Tag extends BaseItem {
+		const oldParentTagId = tag.parent_id;
+
+		// Save new parent id
+		tag.parent_id = parentTagId;
+		tag = await Tag.save(tag, { userSideValidation: true });
```
Please don't mutate data, and ideally only save the properties that have changed - `const newTag = await Tag.save({ id: tag.id, parent_id: parentTagId }, { userSideValidation: true });`

laurent22 commented on ReactNativeClient/lib/models/Tag.js:
```diff
@@ -174,23 +342,131 @@ class Tag extends BaseItem {
+
+		let parentId = '';
+		// Check if the tag is nested using `/` as separator
+		const separator = '/';
+		const i = fullTitle.lastIndexOf(separator);
+		if (i !== -1) {
+			const parentTitle = fullTitle.slice(0,i);
+			tag.title = fullTitle.slice(i + 1);
+
+			// Try to get the parent tag
+			const parentTag = await Tag.loadByTitle(parentTitle);
+			// The second part of the conditions ensures that we do not create a loop
+			// in the tag hierarchy
+			if (parentTag &&
+				!('mainTagId' in options
+					&& (options.mainTagId === parentTag.id
+						|| (await Tag.descendantTagIds(options.mainTagId)).includes(parentTag.id)))
+			) {
+				parentId = parentTag.id;
+			} else {
+				// Create the parent tag if it doesn't exist
+				const parentOpts = {};
+				if ('mainTagId' in options) parentOpts.mainTagId = options.mainTagId;
+				const parentTag = await Tag.saveNested({}, parentTitle, parentOpts);
+				parentId = parentTag.id;
+			}
+		} else {
+			// Tag is not nested so set the title to full title
+			tag.title = fullTitle;
+		}
+
+		// Set parent_id
+		tag.parent_id = parentId;
```
Please save the tag first to a new object to avoid mutations - `const newTag = Object.assign({}, tag)`

laurent22 commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -284,6 +284,70 @@ class Api {
+		if (hasFullTitleField && request.method === 'GET' && !id) {
+			const tags = await this.defaultAction_(BaseModel.TYPE_TAG, request, id, link);
+
+			for (let i = 0; i < tags.length; i++) {
+				tags[i].full_title = Tag.getCachedFullTitle(tags[i].id);
+			}
```
Likewise, please use map/Object.assign to copy the data before modifying it.

laurent22 commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -284,6 +284,70 @@ class Api {
+			return tags;
+		}
+
+		if (hasFullTitleField && request.method === 'GET' && id) {
+			const tag = await this.defaultAction_(BaseModel.TYPE_TAG, request, id, link);
+			tag.full_title = Tag.getCachedFullTitle(tag.id);
+			return tag;
```
Same here

vsimkus commented:
@laurent22 Thanks again. I've made the requested changes. Before, I kept the `note_count` virtual property as it was added before this PR, since I didn't want to mix changes unrelated to nesting in the same PR. 

I've added now a noteCount cache and `Tag.getCachedNoteCount(tagId)` method to retrieve note counts, similar to fullTitle. I couldn't make it without the cache due to async/non-async function call mixing - the retrieval of noteCounts requires several async calls, but the react render function is not async.

laurent22 commented:
Perfect, I have now merged the changes and will release the feature in the next pre-release so that it can be beta tested. Many thanks for this very useful feature @vsimkus which I think many users will appreciate since it was the third most upvoted issue on GitHub!

vsimkus commented:
Glad to know that!

genneko commented:
I've just found this feature available on the head revision. Thank you so much for this enhancement.

I'm not sure if it's appropriate to comment on this here but one thing I noticed is that tags are now displayed in a different order than before. It would be great if they could be sorted in the same way as before.

vsimkus commented:
@genneko Could you elaborate where the tag order had changed? I could take a look.

genneko commented:
@vsimkus Sure. Tags had been sorted in alphabetical or ascii-like order. Now the order is seemingly random. The difference can also be seen on the "Before/After" screenshots in your [comment here](https://github.com/laurent22/joplin/pull/2572#issuecomment-633102823). Thank you.

vsimkus commented:
@genneko Indeed the alphabetical sorting has been removed accidentally during refactoring. Added it back in another PR #3489.

StanczakDominik commented:
Oh wow, now that's just glorious. Thanks for doing this!

genneko commented:
@vsimkus Confirmed the sort order had been restored by #3489. Perfect. Thanks!

laurent22 commented:
@vsimkus, do you know of any remaining bug in the hierarchical tag feature? It seems quite stable so I think it could be part of the next stable release.

fxshpac commented:
One other thing I noticed is extended wait (2+ seconds) needed to bring up the tag dialog by clicking the icon below. It used to be almost instant before this update.

![image](https://user-images.githubusercontent.com/5333310/88140028-8675a500-cc23-11ea-8f92-c00e042e677d.png)

Following are the stats of my joplin dataset just FYI.

![image](https://user-images.githubusercontent.com/5333310/88140174-d05e8b00-cc23-11ea-8230-14e1be95ea43.png)

And following is the version info:

Joplin 1.0.229 (prod, win32)

Client ID: 0c95292a638d44e4b15cc6a5f8eb4f9d
Sync Version: 1
Profile Version: 31
Keychain Supported: No

Revision: d0b3e159 (master)

Besides the symptom above, I also notice that Android version is not supporting nested tags yet. Can we also build an equivalent version for Android for us to try?

vsimkus commented:
@fxshpac Thanks for the detailed report! It is expected to be slower, but ideally the performance should be similar - I will look into it.

@laurent22 I haven't found any additional bugs that weren't already reported here. The issue with the slow performance performance reported above is two-fold: (a) the method calls `Tag.all` that attempts to refresh the `noteCount` cache, which is quite expensive, at least in the current implementation, and (b) it calls `Tag.all` twice - through two different functions.

The issue (b) can be fixed quite simply, so that would cut the load time by two, but it is still quite slow. Regarding (a) there are at least two approaches to handle it:

1. Don't reload the note count cache in `Tag.all()`. This would require first initialising the cache when the application starts through an additional call (maybe before Tag.allWithNotes() call`). Also, I think this PR missed a couple of things to make sure that the note count cache is always up to date: we need to refresh the cache items for the tag and its ancestors when `NoteTag.save()` or `NoteTag.delete()` are called, and when `Tag.delete()` or `Tag.save()` are called. Currently, its only updated on `Tag.addNote()`, `Tag.removeNote()`, and `Tag.all()` calls. I suppose the state of the note count cache could be maintained by adding another event type, emitting it whenever `delete` or `save` is called on `Tag` or `NoteTag`, and handling it in the shared-middleware.
2. Make `Tag.updateCachedNoteCountForIds()` more performant by using a single (although complex/recursive) sql query. Similar to what I've done in the beginning.

Ideally it would be great if the maintaining of the tag note count cache could be somehow centralised.

Let me know what you think.

bela53 commented:
First: Thank you for implementing this useful feature.

Does anybody else experience severe performance issues? With current pre-release 1.0.231, ~ 2800 migrated notes and many pre-existing nested tags, I get a delay of *10 - 15 seconds*, until the the "Tags" dialog for a selected note opens (Ubuntu 20.04). The used desktop pc is an old one (~5 years), though I wonder, what mobile performance would look like.

vsimkus commented:
@bela53 Thanks for the report! I've created a PR for a draft of the fix but it might take a while before its finished up.

thxv3n0lvl commented:
was this reverted? can't use it in the next version

Joplin 1.4.11 (prod, linux)
Revision: 28f7544 (master)

qingbol commented:
why i still don't have this feature ?
![image](https://user-images.githubusercontent.com/45025489/100120645-2ea3b080-2e46-11eb-8a9e-dbbccc799317.png)

Joplin 1.4.12 (prod, darwin)
Sync Version: 2
Profile Version: 34
Keychain Supported: Yes
Revision: 0f0c50a (master)

tessus commented:
It has been reverted due to performance issues. The original author will look into it when he has more time.
